mempool/cat/pool.go

```diff
@@ -0,0 +1,786 @@
+// the order to which transactions are entered. There is no guarantee when CheckTx
+// passes that a transaction has been successfully broadcast to any of its peers.
+//
+// A TTL can be set to remove transactions after a period of time or a number
+// of heights.
+//
+// A cache of rejectedTxs can be set in the mempool config. Transactions that
+// are rejected because of `CheckTx` or other validity checks will be instantly
+// rejected if they are seen again. Committed transactions are also added to
+// this cache. This serves somewhat as replay protection but applications should 
+// implement something more comprehensive
+type TxPool struct {
+	// Immutable fields
+	logger       log.Logger
+	config       *config.MempoolConfig
+	proxyAppConn proxy.AppConnMempool
+	metrics      *mempool.Metrics
+
+	updateMtx            sync.Mutex
+	notifiedTxsAvailable bool
+	txsAvailable         chan struct{} // one value sent per height when mempool is not empty
+	preCheckFn           mempool.PreCheckFunc
+	postCheckFn          mempool.PostCheckFunc
+	height               int64 // the latest height passed to Update
+
+	// Thread-safe cache of rejected transactions for quick look-up
+	rejectedTxCache *LRUTxCache
+	// Thread-safe cache of valid txs that were evicted
+	evictedTxs *EvictedTxCache
+	// Thread-safe list of transactions peers have seen that we have not yet seen
+	seenByPeersSet *SeenTxSet
+
+	// Store of wrapped transactions
+	store *store
+
+	// broadcastCh is an unbuffered channel of new transactions that need to
+	// be broadcasted to peers. Only populated if `broadcast` is enabled
+	broadcastCh      chan types.TxKey
+	broadcastMtx     sync.Mutex
+	txsToBeBroadcast map[types.TxKey]struct{}
+}
+
+// NewTxPool constructs a new, empty content addressable txpool at the specified
+// initial height and using the given config and options.
+func NewTxPool(
+	logger log.Logger,
+	cfg *config.MempoolConfig,
+	proxyAppConn proxy.AppConnMempool,
+	height int64,
+	options ...TxPoolOption,
+) *TxPool {
+	txmp := &TxPool{
+		logger:           logger,
+		config:           cfg,
+		proxyAppConn:     proxyAppConn,
+		metrics:          mempool.NopMetrics(),
+		rejectedTxCache:  NewLRUTxCache(cfg.CacheSize),
+		evictedTxs:       NewEvictedTxCache(evictedTxCacheSize),
+		seenByPeersSet:   NewSeenTxSet(),
+		height:           height,
+		preCheckFn:       func(_ types.Tx) error { return nil },
+		postCheckFn:      func(_ types.Tx, _ *abci.ResponseCheckTx) error { return nil },
+		store:            newStore(),
+		broadcastCh:      make(chan types.TxKey, 1),
+		txsToBeBroadcast: make(map[types.TxKey]struct{}),
+	}
+
+	for _, opt := range options {
+		opt(txmp)
+	}
+
+	return txmp
+}
+
+// WithPreCheck sets a filter for the mempool to reject a transaction if f(tx)
+// returns an error. This is executed before CheckTx. It only applies to the
+// first created block. After that, Update() overwrites the existing value.
```

Comment:
> It only applies to the
// first created block. After that, Update() overwrites the existing value.

What does the first created block mean? You mean f(tx) is applied to all the transaction in the Genesis block?
Also, can you please expand on the second part i.e.,
> After that, Update() overwrites the existing value.

It is not clear immediately.

Reply:
Yes all the txs before height 1 will use this preCheck and postCheck function. Afterwards, the `Update` function includes new preCheck and postCheck functions that will overwrite the ones initially provided in the constructor. Does that make sense? Shall I modify the comments as so (for the record those weren't my comments to begin with)

Reply:
Thanks for the clarification!
> Shall I modify the comments as so (for the record those weren't my comments to begin with)

I'd say yes, that would help the readability of the code